ql/src/java/org/apache/hadoop/hive/ql/ppd/OpProcFactory.java

```diff
@@ -782,6 +790,89 @@ protected ExprWalkerInfo mergeChildrenPred(Node nd, OpWalkerInfo owi,
     }
   }
 
+  protected static Object splitFilter(FilterOperator op,
+      ExprWalkerInfo ewi, OpWalkerInfo owi) throws SemanticException {
+
+    RowSchema inputRS = op.getSchema();
+
+    Map<String, List<ExprNodeDesc>> pushDownPreds = ewi.getFinalCandidates();
+    Map<String, List<ExprNodeDesc>> unPushDownPreds = ewi.getNonFinalCandidates();
+
+    // combine all deterministic predicates into a single expression
+    List<ExprNodeDesc> deterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator1 = pushDownPreds.values().iterator();
+    while (iterator1.hasNext()) {
+      for (ExprNodeDesc pred : iterator1.next()) {
+        deterministicPreds = ExprNodeDescUtils.split(pred, deterministicPreds);
+      }
+    }
+
+    if (deterministicPreds.isEmpty()) {
+      return null;
+    }
+
+    List<ExprNodeDesc> nondeterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator2 = unPushDownPreds.values().iterator();
+    while (iterator2.hasNext()) {
+      for (ExprNodeDesc pred : iterator2.next()) {
+        nondeterministicPreds = ExprNodeDescUtils.split(pred, nondeterministicPreds);
+      }
+    }
+
+    assert !nondeterministicPreds.isEmpty();
+
+    ExprNodeDesc deterministicCondn = ExprNodeDescUtils.mergePredicates(deterministicPreds);
+    ExprNodeDesc nondeterministicCondn = ExprNodeDescUtils.mergePredicates(nondeterministicPreds);
+
+    Operator<FilterDesc> deterministicFilter =
+        OperatorFactory.get(new FilterDesc(deterministicCondn, false), new RowSchema(inputRS.getSignature()));
+
+    deterministicFilter.setChildOperators(new ArrayList<Operator<? extends OperatorDesc>>());
+    deterministicFilter.getChildOperators().add(op);
+
+    List<Operator<? extends OperatorDesc>> originalParents = op
+        .getParentOperators();
+    for (Operator<? extends OperatorDesc> parent : originalParents) {
+      List<Operator<? extends OperatorDesc>> childOperators = parent
+          .getChildOperators();
+      int pos = childOperators.indexOf(op);
+      childOperators.remove(pos);
+      childOperators.add(pos, deterministicFilter);
+
+      int pPos = op.getParentOperators().indexOf(parent);
+      deterministicFilter.getParentOperators().add(pPos, parent);
+    }
+
+    op.getParentOperators().clear();
+    op.getParentOperators().add(deterministicFilter);
+    op.getConf().setPredicate(nondeterministicCondn);
+
+    if (HiveConf.getBoolVar(owi.getParseContext().getConf(),
+        HiveConf.ConfVars.HIVEPPDREMOVEDUPLICATEFILTERS)) {
+      // remove the candidate filter ops
+      for (FilterOperator fop : owi.getCandidateFilterOps()) {
```

Comment:
why do we remove the candidate filter operation here?

Reply:
there is the key to split deterministic predicates and nondeterministic predicates